jmespath-core/src/main/java/io/burt/jmespath/function/MergeFunction.java

```diff
@@ -0,0 +1,43 @@
+package io.burt.jmespath.function;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.LinkedHashMap;
+
+import io.burt.jmespath.Adapter;
+
+public class MergeFunction extends JmesPathFunction {
+  public MergeFunction() {
+    super(1, Integer.MAX_VALUE);
+  }
+
+  @Override
+  protected <T> T internalCall(Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    if (isObjectArray(adapter, arguments)) {
+      Map<String, T> accumulator = new LinkedHashMap<>();
+      for (ExpressionOrValue<T> argument : arguments) {
+        T value = argument.value();
+        for (String property : adapter.getPropertyNames(value)) {
+          accumulator.put(property, adapter.getProperty(value, property));
+        }
```

Comment:
`putObjectProperty` would require all implementations to have a mutable JSON object type, which I think is unfortunate. I also don't really see the benefit of `Iterator<Map.Entry<T,T>>` over `Map<T,T>` (but I'm starting to get over on the `<T,T>` side over `<String,T>`).